### DISCUSSION TITLE
Split `insight`

### DISCUSSION THREAD

**Discussion by mattansb** [easystats/insight#701] [what's your opinion]
Currently `insight` has two types of functions:

1. `insight` functions - for extracting information from R models.
2. `easystats` infrastructure functions - printing, formatting, etc.

In the long run, I think it would make sense to split that last part off into a separate function *or* have it moved to `datawizard`?

@easystats/maintainers Your thoughts?

**Comment by IndrajeetPatil** [comment]
Good idea.

I'd personally prefer if we make a new package that contains all printing and formatting relevant functionality.

**Comment by strengejacke** [comment]
While I in general agree with you, and in particular since we want packages to have a very specific focus, my fear is that we introduce a lot of "self-dependencies". That is, while we don't need to import _other_ packages and can follow a zero-dependency policy, we import dozens of _our_ packages.

**Reply by IndrajeetPatil** [reply]
>  That is, while we don't need to import other packages and can follow a zero-dependency policy, we import dozens of our packages.

True, but note that this is going to happen mostly for our user-facing packages (see, report, etc.), and they don't tend to be concerned too much about dependency management.

But, our low-level packages (insight, datawizard, etc.) target mainly developers, who are concerned about dependencies, and these packages import very few of our own packages.

So further modularization of the ecosystem shouldn't be an issue either for users or for developers.

**Reply by mattansb** [reply]
Many dependencies aren't really an issue when they are all maintained in a coherent matter by a team working together.
So although technically splitting off some functions will add 1 dep, functionally they will be working just the same as before.

And so splitting only has the benefits of being easier to maintain (don't need to update insight just for some change in `format_table()` to affect all down stream packages), and potentially easier to use by other devs.

**Reply by vincentarelbundock** [reply]
If we are just talking about the printing and error/message functions, it seems fine to keep those here. I'm not sure there's enough "meat" there to justify a separate package anyway, since (a) nobody but us is likely to use that, and (b) our release cycle is already insanely complicated, and further fractionalization can only make that worse. 

Probably an unpopular opinion, but as a downstream developer who relies on `easystats` a lot, my absolute (selfish) priorities would be:

1. Limit scope creep and new functions to preserve maintainability in the long run.
2. Stop moving and renaming things.

I've been guilty of both of those in my own contributions to `easystats`, so this is not an indictment of anyone. But I really think we ought to move out of a "break things" mindset, and toward a 1.0 (even if that is still years away IMHO).

**Reply by vincentarelbundock** [reply]
> So further modularization of the ecosystem shouldn't be an issue either for users or for developers.

Disagree. I would have to make a bunch of changes to `modelsummary` and `marginaleffects`, import a new dependency, and coordinate release cycles with another upstream package.

Honestly, I'd probably just bite the bullet and write my own utility functions rather than import yet another package.

**Reply by IndrajeetPatil** [reply]
> Disagree. I would have to make a bunch of changes to modelsummary and marginaleffects, import a new dependency, and coordinate release cycles with another upstream package.

Huh?

modelsummary and marginaleffects import insight, which itself will import this new package. So you won't need to do anything.

**Reply by vincentarelbundock** [reply]
Granted, it's not a *ton* of work, but I would still have to:

1. Figure out which functions are moved
2. Search and replace all appropriate `insight::fun()` calls (I don't import the full namespace).
3. Make two well-timed CRAN releases.
4. Hope that none of my users update `insight` without updating my packages (assuming `insight` won't reexport everything)

Also, we lose the nice no recursive 0-dep pitch for `insight`.

Are there plans to add a bunch more `insight`-irrelevant functions? How many are there now?

To be clear, I'm not saying this is a huge deal. But while I see the costs, I don't quite see the gains. This just feels like busywork that introduces yet more instability.

**Reply by mattansb** [reply]
These are sort term pains, I think - I am wondering if there are long term gains to be had...

**Reply by mattansb** [reply]
For consistency, I do think data formatting stuff should just be in `datawizard` 🤷‍♂️

**Reply by vincentarelbundock** [reply]
I slept badly, so maybe I'm just feeling argumentative 🤣🤐

**Reply by strengejacke** [reply]
>  I do think data formatting stuff should just be in datawizard

What other formatting stuff do we have?

**Reply by strengejacke** [reply]
I think most critical will be `format_alert()`, `format_message()`, `format_warning()` and `format_error()`. We could move all other `format_*()` functions without needing to import them in _insight_ nor reexporting and it won't break any external package, I guess.

That would mean we would have two lowest-level packages. If we want to move the format-messages as well, this would break a lot (unless we re-export).

So we just could move all formatters (except messages) to _datawizard_, and nothing will change for other developers, unless there are some recent changes / packages I'm not aware of. However, we also have the print/display-generics, and `format_table()` and `export_table()`. ❓

**Reply by bwiernik** [reply]
I don't see that much benefit from splitting off the formatting functions. The only real upside I think would be if some developer wanted the formatting functions but not the other *insight* features. 

The formatting functions have r been in *insight* for a long time, and they do make some thematic sense to me (*insight* is utility functions for pulling stuff out of models and manipulating them).

I think I agree with @vincentarelbundock here

**Reply by DominiqueMakowski** [reply]
I also think we should leave it as-is. 

If we had to re-start easystats today, we would have done things differently, but what is important now is that it's stable, it's okay if things are not perfectly aligned within the perfect easystats conceptual entity.

(also, 10 packages is a nice number👌)

**Reply by bwiernik** [reply]
> also, 10 packages is a nice number👌

We'd need at least 4 new packages to keep the sticker wall looking pretty
